src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs

```diff
@@ -1154,12 +1157,52 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
                     nameof(expression));
             }
 
+            var inputTypeString = GetInputTypeString(inputType);
             var tagBuilder = new TagBuilder("input");
             tagBuilder.TagRenderMode = TagRenderMode.SelfClosing;
             tagBuilder.MergeAttributes(htmlAttributes);
-            tagBuilder.MergeAttribute("type", GetInputTypeString(inputType));
+            tagBuilder.MergeAttribute("type", inputTypeString);
             tagBuilder.MergeAttribute("name", fullName, replaceExisting: true);
 
+            switch (inputType)
+            {
+                case InputType.Text:
+                    object suppliedTypeString;
+                    if (htmlAttributes != null &&
+                        htmlAttributes.TryGetValue("type", out suppliedTypeString) &&
+                        suppliedTypeString != null)
+                    {
+                        if (_placeholderInputTypes.Contains(suppliedTypeString))
+                        {
+                            goto case InputType.Password;
+                        }
+                    }
+                    else if (inputTypeString == "text")
+                    {
+                        goto case InputType.Password;
+                    }
+                    break;
+
+                case InputType.Password:
+                    modelExplorer = modelExplorer ?? ExpressionMetadataProvider.FromStringExpression(expression, viewContext.ViewData, _metadataProvider);
+                    if (modelExplorer.Metadata.Placeholder != null)
```

Comment:
What I meant in [this comment](https://github.com/aspnet/Mvc/pull/3788#discussion_r48499078) is that calling `.Placeholder` invokes some user code which might call back into localization, pull this out into a variable.